lib/Yield/Support/GraphLayout/SugiyamaStyle/GraphPreparation.cpp

```diff
@@ -42,39 +43,66 @@ static SelfLoopContainer extractSelfLoops(InternalGraph &Graph) {
   return Result;
 }
 
-/// To simplify the ranking algorithms, if there's more than one entry point,
-/// an artificial entry node is added. This new node has a single edge per
-/// real entry point.
-static void
-ensureSingleEntry(InternalGraph &Graph, RankContainer *MaybeRanks = nullptr) {
-  auto EntryNodes = entryPoints(&Graph);
-  revng_assert(!EntryNodes.empty());
-
-  if (EntryNodes.size() == 1) {
-    // If there's only a single entry point, make sure it's set.
-    Graph.setEntryNode(EntryNodes.front());
-  } else {
-    // If there's more than one, add a new virtual node with all the real
-    // entry nodes as its direct successors.
-
-    // BUT if the currently set entry node is already virtual, remove it first:
-    // this prevents the possibility of chaining virtual entry nodes when this
-    // function is invoked on a slightly-modified graph multiple times.
-    if (Graph.getEntryNode() != nullptr) {
-      if (Graph.getEntryNode()->IsVirtual) {
-        Graph.removeNode(Graph.getEntryNode());
-        if (MaybeRanks != nullptr)
-          MaybeRanks->erase(Graph.getEntryNode());
+/// Adds a virtual node to the graph to serve as an entry point, while
+/// also cleanly removing it when this object is being destroyed.
+class VirtualEntryPointGuard {
+private:
+  InternalGraph &Graph;
+  InternalNode &Node;
```

Comment:
So this is the virtual entry point right? Can you call it `VirtualEntryPoint` please?

Reply:
I went with `VirtualEntry` (80 character columns are brutal).

Closing this, feel free to reopen if you don't like it.